src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,566 @@
+		GetRangeTableEntriesFromJoinTree(joinExpr->larg, rangeTableList,
+										 joinRangeTableEntries);
+		GetRangeTableEntriesFromJoinTree(joinExpr->rarg, rangeTableList,
+										 joinRangeTableEntries);
+	}
+	else if (IsA(joinNode, RangeTblRef))
+	{
+		int rangeTableIndex = ((RangeTblRef *) joinNode)->rtindex;
+		RangeTblEntry *rte = rt_fetch(rangeTableIndex, rangeTableList);
+		*joinRangeTableEntries = lappend(*joinRangeTableEntries, rte);
+	}
+	else
+	{
+		pg_unreachable();
+	}
+}
+
+
+/*
+ * GetNextRTEToConvertToSubquery returns the range table entry
+ * which should be converted to a subquery. It considers the local join policy
+ * and result relation.
+ */
+static RTEToSubqueryConverterReference *
+GetNextRTEToConvertToSubquery(FromExpr *joinTree,
+							  RTEToSubqueryConverterContext *rteToSubqueryConverterContext,
+							  PlannerRestrictionContext *plannerRestrictionContext, Oid
+							  resultRelationId)
+{
+	RTEToSubqueryConverterReference *localRTECandidate = NULL;
+	RTEToSubqueryConverterReference *distributedRTECandidate = NULL;
+	if (!FillLocalAndDistributedRTECandidates(rteToSubqueryConverterContext,
+											  &localRTECandidate,
+											  &distributedRTECandidate))
+	{
+		return NULL;
+	}
+
+	if (OidIsValid(resultRelationId))
+	{
+		if (resultRelationId == GetRTEToSubqueryConverterReferenceRelId(
+				localRTECandidate))
+		{
+			return distributedRTECandidate;
+		}
+		if (resultRelationId == GetRTEToSubqueryConverterReferenceRelId(
+				distributedRTECandidate))
+		{
+			return localRTECandidate;
+		}
+	}
+
+	if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_PREFER_LOCAL)
+	{
+		return localRTECandidate;
+	}
+	else if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_PREFER_DISTRIBUTED)
+	{
+		if (distributedRTECandidate)
+		{
+			return distributedRTECandidate;
+		}
+		else
+		{
+			return localRTECandidate;
+		}
+	}
+	else
+	{
+		if (ShouldConvertDistributedTable(joinTree, distributedRTECandidate))
+		{
+			return distributedRTECandidate;
+		}
+		else
+		{
+			return localRTECandidate;
+		}
+	}
+}
+
+
+/*
+ * FillLocalAndDistributedRTECandidates fills the local and distributed RTE candidates.
+ * It returns true if we should continue converting tables to subqueries.
+ */
+static bool
+FillLocalAndDistributedRTECandidates(
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext,
+	RTEToSubqueryConverterReference **localRTECandidate,
+	RTEToSubqueryConverterReference **
+	distributedRTECandidate)
+{
+	if (list_length(rteToSubqueryConverterContext->localTableList) > 0)
+	{
+		*localRTECandidate = linitial(rteToSubqueryConverterContext->localTableList);
+	}
+	if (*localRTECandidate == NULL)
+	{
+		return false;
+	}
+
+	if (list_length(rteToSubqueryConverterContext->distributedTableList) > 0)
+	{
+		*distributedRTECandidate = linitial(
+			rteToSubqueryConverterContext->distributedTableList);
+	}
+	return *distributedRTECandidate != NULL ||
+		   rteToSubqueryConverterContext->hasSubqueryRTE;
```

Comment:
what does the hasSubqueryRTE check mean here? it seems out of place. Should the caller be checking this?

Reply:
I simplified this part as well, so now we don't have hasSubqueryRTE field